multilib-fpus.py

```diff
@@ -0,0 +1,214 @@
+#!/usr/bin/env python3
+
+"""Auto-generate implications between -mfpu options for multilib.yaml.in.
+
+Each FPU name that clang knows about is mapped to all the FPU names
+that clang considers to be a subset of it (determined by extracting
+the cc1 command from `clang -###` for each one and looking at the set
+of -target-feature options).
+
+An exception is that we don't consider any hardware FP configuration
+to be compatible with -mfpu=none. It would work in most cases to
+cross-call between code compiled for an FPU or no FPU, if you were
+using the soft float ABI. But it wouldn't work in all cases: setjmp
+needs to know whether to save FP registers in the jmp_buf, so a
+non-FPU-aware setjmp would not behave correctly if linked into an
+otherwise FPU-using application. Similarly for exception unwinding. So
+we don't permit selecting an -mfpu=none library as a fallback for any
+hard-FP library.
+
+However, it's fine for ABI purposes to mix code compiled for 16 and 32
+d-registers, because the extra 16 d-registers are caller-saved, so
+setjmp and exceptions need not preserve them. Interrupt handlers would
+have to preserve them, but our libraries don't define any.
+"""
+
+import argparse
+import json
+import os
+import shlex
+import subprocess
+
+
+def get_fpu_list(args):
+    """Extract the list of FPUs from ARMTargetParser.def.
+
+    Strategy: the intended use of ARMTargetParser.def in the actual
+    LLVM build is to run it through the C preprocessor with whatever
+    #defines will generate the output you want. So the most reliable
+    way to get what _we_ want is to do exactly the same.
+
+    The output format I've chosen is actually JSON, because that's
+    close enough to C-like syntax that you can generate it easily
+    using cpp features like stringification, and also convenient for
+    Python to consume afterwards.
+    """
+
+    command = [
+        args.clang,
+        # For this purpose we're only interested in the calls to the
+        # ARM_FPU macro in the input file, and want the first and
+        # third argument of each of those.
+        #
+        # The first argument will be a string literal giving the FPU
+        # name, which we copy into the JSON output still as a string
+        # literal.
+        #
+        # The third argument indicates the general FPU category, which
+        # we need so as to exclude FPUVersion::NONE. That is not
+        # already a string literal, so we stringify it in the
+        # preprocessor to make it legal JSON.
+        "-DARM_FPU(a,b,c,...)=[a,#c],",
```

Comment:
I made this comment before but GitHub seems to have lost it:
```suggestion
        "-DARM_FPU(name,kind,version,...)=[name,#version],",
```